lib/gc/Transforms/VerifyTargetDescription.cpp

```diff
@@ -0,0 +1,115 @@
+
+namespace mlir {
+namespace gc {
+#define GEN_PASS_DEF_VERIFYTARGETDESCRIPTION
+#include "gc/Transforms/Passes.h.inc"
+
+namespace {
+
+static LogicalResult verifyCPUTargetDescription(RewriterBase &rewriter,
+                                                Operation *op) {
+  CPUTargetDescriptionAnalysis cpuTargetDesc(op);
+  Location loc = op->getLoc();
+
+  // Check if the num_threads is existed and greater than 0
+  std::optional<Attribute> numThreadsAttr =
+      cpuTargetDesc.getPropertyValue(CPUTargetDescriptionAnalysis::kNumThreads);
+  if (numThreadsAttr) {
+    if (!isa<IntegerAttr>(*numThreadsAttr) ||
+        cpuTargetDesc.getNumThreads() < 1) {
+      mlir::emitError(loc)
+          << "num_threads must be a greater than 0 integer, but get "
+          << *numThreadsAttr;
+      return failure();
+    }
+  }
+
+  // Check if the L1 cache size is existed and greater than 0
+  std::optional<Attribute> l1CacheSizeAttr = cpuTargetDesc.getPropertyValue(
+      CPUTargetDescriptionAnalysis::kL1CacheSize);
+  if (l1CacheSizeAttr) {
+    if (!isa<IntegerAttr>(*l1CacheSizeAttr) ||
+        cpuTargetDesc.getCacheSize(1) < 1) {
+      mlir::emitError(loc)
+          << "L1_cache_size_in_bytes must be a greater than 0 integer, but get "
+          << *l1CacheSizeAttr;
+      return failure();
+    }
+  }
+
+  // Check if the L2 cache size is existed and greater than 0
+  std::optional<Attribute> l2CacheSizeAttr = cpuTargetDesc.getPropertyValue(
+      CPUTargetDescriptionAnalysis::kL2CacheSize);
+  if (l2CacheSizeAttr) {
+    if (!isa<IntegerAttr>(*l2CacheSizeAttr) ||
+        cpuTargetDesc.getCacheSize(2) < 1) {
+      mlir::emitError(loc)
+          << "L2_cache_size_in_bytes must be a greater than 0 integer, but get "
+          << *l2CacheSizeAttr;
+      return failure();
+    }
+  }
+
+  // Check if the L3 cache size is existed and greater than 0
+  std::optional<Attribute> l3CacheSizeAttr = cpuTargetDesc.getPropertyValue(
+      CPUTargetDescriptionAnalysis::kL3CacheSize);
+  if (l3CacheSizeAttr) {
+    if (!isa<IntegerAttr>(*l3CacheSizeAttr) ||
+        cpuTargetDesc.getCacheSize(3) < 1) {
+      mlir::emitError(loc)
+          << "L3_cache_size_in_bytes must be a greater than 0 integer, but get "
+          << *l3CacheSizeAttr;
+      return failure();
+    }
+  }
+
+  // Check if the max_vector_width is existed and greater than 0
+  std::optional<Attribute> maxVectorWidthAttr = cpuTargetDesc.getPropertyValue(
+      CPUTargetDescriptionAnalysis::kMaxVectorWidth);
+  if (maxVectorWidthAttr) {
+    if (!isa<IntegerAttr>(*maxVectorWidthAttr) ||
+        cpuTargetDesc.getMaxVectorWidth() < 1) {
+      mlir::emitError(loc)
+          << "max_vector_width must be a greater than 0 integer, but get "
+          << *maxVectorWidthAttr;
+      return failure();
+    }
+  }
+  return success();
+}
+
+class VerifyTargetDescription
+    : public impl::VerifyTargetDescriptionBase<VerifyTargetDescription> {
+  using Base::Base;
```

Comment:
Will it be placed at the first pass of the whole pipeline? But I dont find it anywhere in pipeline at present.

Reply:
Sorry for the missing. I have added it to the pipeline.